src/Query.tsx

```diff
@@ -147,86 +184,92 @@ export default class Query<TData = any, TVariables = OperationVariables> extends
-    this.client = props.client || context.client;
-    invariant(
-      !!this.client,
-      `Could not find "client" in the context of Query or as passed props. Wrap the root component in an <ApolloProvider>`,
-    );
-    this.initializeQueryObservable(props);
+    if (nextProps.skip) return null;
+
+    prevState.evictData = false;
+
+    // remove the queryObservable so cDU will have to create a new one
+    if (nextProps.client !== prevState.props.client || nextProps.query !== prevState.props.query) {
+      prevState.evictData = true;
+      prevState.queryObservable = null;
+    }
+
+    // update the ObservableQuery
+    prevState.queryObservable = updateQuery(nextProps, prevState);
+
+    return prevState;
+  }
+  constructor(props: InnerQueryProps<TData, TVariables>) {
+    super(props);
+    this.state = {
+      queryObservable: initializeQueryObservable(props),
+      props,
+    } as any;
   }
 
   // For server-side rendering (see getDataFromTree.ts)
   fetchData(): Promise<ApolloQueryResult<any>> | boolean {
     if (this.props.skip) return false;
     // pull off react options
-    const { children, ssr, displayName, skip, client, ...opts } = this.props;
+    const { children, ssr, displayName, skip, client, asyncMode, ...opts } = this.props;
 
     let { fetchPolicy } = opts;
     if (ssr === false) return false;
     if (fetchPolicy === 'network-only' || fetchPolicy === 'cache-and-network') {
       fetchPolicy = 'cache-first'; // ignore force fetch in SSR;
     }
 
-    const observable = this.client.watchQuery({
+    const observable = this.props.client.watchQuery({
       ...opts,
       fetchPolicy,
     });
-    const result = this.queryObservable!.currentResult();
+    const result = this.state.queryObservable!.currentResult();
 
     return result.loading ? observable.result() : false;
   }
 
   componentDidMount() {
     this.hasMounted = true;
     if (this.props.skip) return;
-    this.startQuerySubscription();
+    this.startQuerySubscription(this.state.queryObservable);
     if (this.refetcherQueue) {
       const { args, resolve, reject } = this.refetcherQueue;
-      this.queryObservable!
-        .refetch(args)
+      this.state
+        .queryObservable!.refetch(args)
         .then(resolve)
         .catch(reject);
     }
   }
 
-  componentWillReceiveProps(nextProps: QueryProps<TData, TVariables>, nextContext: QueryContext) {
+  componentDidUpdate(prevProps: InnerQueryProps<TData, TVariables>) {
     // the next render wants to skip
-    if (nextProps.skip && !this.props.skip) {
+    if (this.props.skip && !prevProps.skip) {
       this.removeQuerySubscription();
       return;
     }
 
-    const { client } = nextProps;
-    if (
-      shallowEqual(this.props, nextProps) &&
-      (this.client === client || this.client === nextContext.client)
-    ) {
-      return;
-    }
+    // if there are no changes to the props, don't do anything state wise
+    if (shallowEqual(this.props, prevProps)) return null;
```

Comment:
Again, this short-circuit is unnecessary, given that everything below is guarded with more specific conditionals.

Plus, in the world of render-props (where the prevailing trend is to use inline arrow functions), the `children` prop is _never_ going to equal its prior value. Might as well save some cycles and cognitive overhead.